src/features/common/data/indexed-db.ts

```diff
@@ -1,32 +1,89 @@
-import { DBSchema, IDBPDatabase, openDB } from 'idb'
+import { DBSchema, IDBPDatabase, IDBPTransaction, openDB, StoreNames } from 'idb'
 import { AppSpecVersion } from '@/features/abi-methods/data/types'
+import { ApplicationId } from '@/features/applications/data/types'
 
-interface LoraDBSchema extends DBSchema {
+interface LoraDBSchemaV1 extends DBSchema {
   'applications-app-specs': {
     key: string
     value: AppSpecVersion[]
   }
 }
 
+export type ContractEntity = {
+  applicationId: ApplicationId
+  displayName: string
+  appSpecVersions: AppSpecVersion[]
+  lastModified: number
+}
+
+interface LoraDBSchemaV2 extends DBSchema {
+  contracts: {
+    key: number
+    value: ContractEntity
+  }
+}
+
 export async function getDbConnection(networkId: string) {
-  return await openDB<LoraDBSchema>(networkId, 1, {
-    upgrade(db, oldVersion, newVersion) {
-      const migrationsToRun = dbMigrations.slice(oldVersion, newVersion ?? undefined)
-      migrationsToRun.forEach((migration) => {
-        migration(db)
-      })
+  return await openDB<LoraDBSchemaV2>(networkId, 2, {
+    async upgrade(db, oldVersion, newVersion, transaction) {
+      // Casting to DBSchema to trick TypeScript
+      const anyDb = db as unknown as IDBPDatabase<DBSchema>
+      const anyTransaction = transaction as unknown as IDBPTransaction<DBSchema, StoreNames<DBSchema>[], 'versionchange'>
+      try {
+        const migrationsToRun = dbMigrations.slice(oldVersion, newVersion ?? undefined)
+        for (const migration of migrationsToRun) {
+          await migration(anyDb, anyTransaction)
+        }
+      } catch (e) {
+        // Need to abort the transaction here, so that the database version doesn't get updated to the new version
+        transaction.abort()
+      }
     },
   })
 }
 
-export let dbConnection: Promise<IDBPDatabase<LoraDBSchema>> | undefined = undefined
+export let dbConnection: Promise<IDBPDatabase<LoraDBSchemaV2>> | undefined = undefined
 
 export const updateDbConnection = (networkId: string) => {
   dbConnection = getDbConnection(networkId)
 }
 
 const dbMigrations = [
-  (db: IDBPDatabase<LoraDBSchema>) => {
-    db.createObjectStore('applications-app-specs')
+  async (db: IDBPDatabase<DBSchema>) => {
+    const v1Db = db as unknown as IDBPDatabase<LoraDBSchemaV1>
+    v1Db.createObjectStore('applications-app-specs')
+  },
+  async (db: IDBPDatabase<DBSchema>, transaction: IDBPTransaction<DBSchema, StoreNames<DBSchema>[], 'versionchange'>) => {
```

Comment:
It's good to have run a migration, however I don't think we need to do do it at this stage. Feels like it'll keep the code simpler to just make the breaking change?

We can probably apply a run that until we release this publicly, we are ok to simply make breaking schema changes. Thoughts?

Alternatively we can apply schema migrations and just clean them up prior to releasing?

Reply:
I don't mind applying breaking changes 😄 . I think the migration here is good for establishing a pattern.

Reply:
Yeah agreed. Maybe let's run the migrations as a way of formalising the process and then cleanup any migration scripts prior to the public release, so we don't have that additional code.

Reply:
Also just realised the db and table is created by default, so we'll need to make sure we account for that.